llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -3949,6 +3949,40 @@ SDValue DAGCombiner::visitSUB(SDNode *N) {
       if (SDValue Result = TLI.expandABS(N1.getNode(), DAG, true))
         return Result;
 
+    // Similar to the previous rule, but this time targeting an expanded abs.
+    // (sub 0, (max X, (sub 0, X))) --> (min X, (sub 0, X))
+    // as well as
+    // (sub 0, (min X, (sub 0, X))) --> (max X, (sub 0, X))
+    // Note that these two are applicable to both signed and unsigned min/max.
+    SDValue X;
+    SDValue S0;
+    auto NegPat = m_AllOf(m_Neg(m_Deferred(X)), m_Value(S0));
+    if (LegalOperations &&
```

Comment:
Technically you should check the max is legal, but I doubt in practice the legality of min is different than max

Reply:
> I doubt in practice the legality of min is different than max

This, and even if max is illegal, I think it'll be expanded (while min is not), which makes this transformation even more profitable.

Reply:
It depends in which combiner phase, eventually only legal operations can be emitted

Reply:
Ah, if you're referring to the max generated by this rule, I already checked its legality (through `NewOpc`) below.